plugins/auth-backend/src/providers/okta/provider.ts

```diff
@@ -0,0 +1,208 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import express from 'express';
+import { OAuthProvider } from '../../lib/OAuthProvider';
+import { Strategy as OktaStrategy } from 'passport-okta-oauth';
+import passport from 'passport';
+import {
+  executeFrameHandlerStrategy,
+  executeRedirectStrategy,
+  executeRefreshTokenStrategy,
+  makeProfileInfo,
+  executeFetchUserProfileStrategy,
+} from '../../lib/PassportStrategyHelper';
+import {
+  OAuthProviderHandlers,
+  RedirectInfo,
+  AuthProviderConfig,
+  EnvironmentProviderConfig,
+  OAuthProviderOptions,
+  OAuthProviderConfig,
+  OAuthResponse,
+  PassportDoneCallback,
+} from '../types';
+import {
+  EnvironmentHandler,
+  EnvironmentHandlers,
+} from '../../lib/EnvironmentHandler';
+import { Logger } from 'winston';
+import { StateStore } from 'passport-oauth2';
+import { TokenIssuer } from '../../identity';
+
+type PrivateInfo = {
+  refreshToken: string;
+};
+
+export class OktaAuthProvider implements OAuthProviderHandlers {
+
+  private readonly _strategy: any;
+
+  /** 
+   * Due to passport-okta-oauth forcing options.state = true,
+   * passport-oauth2 requires express-session to be installed
+   * so that the 'state' parameter of the oauth2 flow can be stored.
+   * This implementation of StateStore matches the NullStore found within
+   * passport-oauth2, which is the StateStore implementation used when options.state = false,
+   * allowing us to avoid using express-session in order to integrate with Okta.
+  */
+  private _store: StateStore = {
+    store({}, cb: any) {
+      cb(null, null);
+    },
+    verify({}, {}, cb: any) {
+      cb(null, true);
+    },
+  }
+
+  constructor(options: OAuthProviderOptions) {
+    this._strategy =  new OktaStrategy({
+      passReqToCallback: false as true,
+      ...options,
+      store: this._store,
+      response_type: 'code',
+    }, (
+      accessToken: any,
+      refreshToken: any,
+      params: any,
+      rawProfile: passport.Profile,
+      done: PassportDoneCallback<OAuthResponse, PrivateInfo>,
+    ) => {
+      const profile = makeProfileInfo(rawProfile, params.id_token);
+
+      done(
+        undefined,
+          {
+            providerInfo: {
+              idToken: params.id_token,
+              accessToken,
+              scope: params.scope,
+              expiresInSeconds: params.expires_in,
+            },
+            profile,
+          },
+          {
+            refreshToken,
+          },
+      )
+    });
+  }
+
+  async start(
+    req: express.Request,
+    options: Record<string, string>
+  ): Promise<RedirectInfo> {
+    return await executeRedirectStrategy(req, this._strategy, options);
```

Comment:
Wonder if we should use `param: consent` here to avoid the instant popup and hide. UX question though and easy to tweak later.

Reply:
I've added the prompt: consent and accessType: offline to these options